xun/functions/driver/dask.py

```diff
@@ -1,52 +1,127 @@
+from .. import graph as graph_helpers
 from .driver import Driver
-import dask
+import asyncio
+import functools
 import logging
 import networkx as nx
 
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
 
 
 class DaskDriverError(Exception):
     pass
 
 
-def compute_proxy(node, dependencies, func, store_accessor):
-    if store_accessor.completed(node):
-        return store_accessor.load_result(node)
+class Dask(Driver):
+    def __init__(self, client):
+        self.client = client
+
+    def _exec(self, graph, entry_call, function_images, store_accessor):
+        assert nx.is_directed_acyclic_graph(graph)
+        scheduler = DaskSchedule(self.client, function_images, store_accessor)
+        return scheduler(entry_call, graph)
 
+
+def compute_proxy(store_accessor, func, node):
     args, kwargs = store_accessor.resolve_call_args(node)
-    try:
-        result = func(*args, **kwargs)
-    except Exception as e:
-        raise DaskDriverError(f'task {node} failed') from e
-    except:
-        raise ValueError('???')
+    result = func(*args, **kwargs)
     store_accessor.store_result(node, result)
     return node
 
 
-class Dask(Driver):
-    def __init__(self, client):
+class DaskSchedule:
+    """
+    Traverse the call graph and submit jobs using an async implementation of
+    Kahn's algorithm [1].
+
+    .. [1] https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+    """
+
+    def __init__(self,
+                 client,
+                 function_images,
+                 store_accessor):
         self.client = client
+        self.function_images = function_images
+        self.store_accessor = store_accessor
 
-    def _exec(self, graph, entry_call, function_images, store_accessor):
-        assert nx.is_directed_acyclic_graph(graph)
+    def __call__(self, entry_call, graph):
+        self.client.sync(self.run, graph)
+        return self.store_accessor.load_result(entry_call)
+
+    async def run(self, graph):
+        atomic_graph = GraphLock(graph)
+        queue = asyncio.Queue()
+
+        consumer = asyncio.ensure_future(self.consume_tasks(atomic_graph,
```

Comment:
Comment for multiple places. I assume that ensure_future is used because we still support python 3.6? Based on the asyncio-task documentation we should use a different function:
https://docs.python.org/3/library/asyncio-task.html#creating-tasks

Should we make sure, that when we'll get out of Python 3.6, we would update all of this?

Reply:
Consider checking for cycles?